pkg/webhook/core.oam.dev/v1alpha2/componentdefinition/validating_handler.go

```diff
@@ -0,0 +1,137 @@
+	// Validators validate objects
+	Validators []ComponentDefValidator
+}
+
+// ComponentDefValidator validate component definition
+type ComponentDefValidator interface {
+	Validate(context.Context, v1beta1.ComponentDefinition) error
+}
+
+// ComponentDefValidatorFn implements ComponentDefValidator
+type ComponentDefValidatorFn func(context.Context, v1beta1.ComponentDefinition) error
+
+// Validate implements ComponentDefValidator method
+func (fn ComponentDefValidatorFn) Validate(ctx context.Context, cd v1beta1.ComponentDefinition) error {
+	return fn(ctx, cd)
+}
+
+var _ admission.Handler = &ValidatingHandler{}
+
+// Handle validate component definition
+func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &v1beta1.ComponentDefinition{}
+	if req.Resource.String() != componentDefGVR.String() {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("expect resource to be %s", componentDefGVR))
+	}
+
+	if req.Operation == admissionv1beta1.Create || req.Operation == admissionv1beta1.Update {
+		err := h.Decoder.Decode(req, obj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if obj.Spec.Workload.Type == "" && obj.Spec.Workload.Definition == (common.WorkloadGVK{}) {
+			errMsg := fmt.Sprintf("neither the type nor the definition of the workload field in the ComponentDefinition %s can be empty", obj.Name)
+			return admission.Denied(errMsg)
+		}
+
+		if obj.Spec.Workload.Type != "" && obj.Spec.Workload.Definition != (common.WorkloadGVK{}) {
+			defRef, err := util.ConvertWorkloadGVK2Definition(h.Mapper, obj.Spec.Workload.Definition)
+			if err != nil {
+				return admission.Denied(err.Error())
+			}
+			if defRef.Name != obj.Spec.Workload.Type {
+				return admission.Denied("the type and the definition of the workload field should represent the same workload")
+			}
+		}
+
+		for _, validator := range h.Validators {
+			if err := validator.Validate(ctx, *obj); err != nil {
+				return admission.Denied(err.Error())
+			}
+		}
+	}
+	return admission.ValidationResponse(true, "")
+}
+
+var _ admission.DecoderInjector = &ValidatingHandler{}
+
+// InjectDecoder injects the decoder into the ValidatingHandler
+func (h *ValidatingHandler) InjectDecoder(d *admission.Decoder) error {
+	h.Decoder = d
+	return nil
+}
+
+// RegisterValidatingHandler will register TraitDefinition validation to webhook
+func RegisterValidatingHandler(mgr manager.Manager, args controller.Args) {
+	server := mgr.GetWebhookServer()
+	server.Register("/validating-core-oam-dev-v1beta1-componentdefinitions", &webhook.Admission{Handler: &ValidatingHandler{
+		Mapper: args.DiscoveryMapper,
+		Validators: []ComponentDefValidator{
+			ComponentDefValidatorFn(ValidateDefinitionReference),
+		},
+	}})
+}
+
+// ValidateDefinitionReference validates whether the component definition is valid
+func ValidateDefinitionReference(_ context.Context, cd v1beta1.ComponentDefinition) error {
```

Comment:
The function name looks not relevant to its logic.

Reply:
got it, i will rename this function name